solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -1818,7 +1820,7 @@ private void getDocListNC(QueryResult qr, QueryCommand cmd) throws IOException {
     int[] ids;
     float[] scores;
 
-    boolean needScores = (cmd.getFlags() & GET_SCORES) != 0;
+    boolean needScores = ((cmd.getFlags() & GET_SCORES) != 0) && (cmd.getLen() != 0);
```

Comment:
Arguably, the caller shouldn't be applying GET_SCORES in the first place -- which is QueryComponent.  I suggested this idea yesterday; do you not like it?

Reply:
I just haven't gotten back to doing that change yet. I will make another iteration.

Reply:
Continuing here https://github.com/apache/solr/pull/2221#discussion_r1471409812